File: 0015-3sum/0015-3sum.java
class Solution {
    public void twoSum(int nums[], int ind, int target, List<List<Integer>> ans) {
        int i = ind;
        int j = nums.length - 1;
 
        while(i < j)
        {
            int b = nums[i];
            int c = nums[j];
            if(b + c == target)
            {
                List<Integer> lst = new ArrayList<>(); 
                lst.add(-target);
                lst.add(b);
                lst.add(c);
                ans.add(lst);
                i ++;
                while(i < j && nums[i] == nums[i - 1])
                    i ++;
                j --;
                while(i < j && nums[j] == nums[j + 1])
                    j --;
            }
            else if(b + c < target)
                i ++;
            else
                j --;
        }
    }
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
 
        for(int i = 0 ; i < nums.length - 2; i ++)
        {
            if(i == 0 || nums[i] != nums[i - 1])
            {
                int a = nums[i];
                twoSum(nums, i + 1, -a, ans);
            }
        }
 
        return ans;
    }
}

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
       //optimal
         vector<vector<int>> ans;
        int n = nums.size();

        // Sort the array to apply the two-pointer approach
        sort(nums.begin(), nums.end());

        // Iterate through the array
        for (int i = 0; i < n; i++) {
            // Skip duplicates for the first element
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            // Use two pointers to find pairs that sum to -nums[i]
            int j = i + 1;
            int k = n - 1;

            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];

                if (sum < 0) {
                    // Move the left pointer to the right to increase the sum
                    j++;
                }
                else if (sum > 0) {
                    // Move the right pointer to the left to decrease the sum
                    k--;
                }
                else {
                    // Found a valid triplet
                    ans.push_back({nums[i], nums[j], nums[k]});

                    // Move both pointers to avoid duplicates
                    j++;
                    k--;

                    // Skip duplicates for the second element
                    while (j < k && nums[j] == nums[j - 1]) j++;
                    // Skip duplicates for the third element
                    while (j < k && nums[k] == nums[k + 1]) k--;
                }
            }
        }

        return ans;
    }
};
